### PR TITLE
 LIVY-617: Livy session leak on Yarn when creating session duplicated names

### PR DESCRIPTION
**Pull request by shanyu**
## What changes were proposed in this pull request?
When creating a session with duplicated name, instead of throw exception in SessionManager.register() method, we should stop the session. Otherwise the session driver process will keep running and end up creating a leaked Yarn application.

https://issues.apache.org/jira/browse/LIVY-617

## How was this patch tested?

This is just a simple fix and verified with manual end to end test.

### CODE REVIEW

**Comment by jerryshao** [comment]
Reopen the PR to retest it.

**Comment by jerryshao** [comment]
Thanks @shanyu for the fix. Would you please add an UT about your fix. Thanks!

**Comment by shanyu** [comment]
@jerryshao UT added. Thanks for reviewing it.

File: server/src/test/scala/org/apache/livy/sessions/MockSession.scala [hunk 1]
```diff
@@ -27,7 +27,8 @@ class MockSession(id: Int, owner: String, conf: LivyConf, name: Option[String] =
 
   override def start(): Unit = ()
 
-  override protected def stopSession(): Unit = ()
+  var stopped = false
+  override protected def stopSession(): Unit = {stopped = true}
```
Comment (jerryshao) [inline server/src/test/scala/org/apache/livy/sessions/MockSession.scala]: nit: change the style to:

```
 = {
stopped = true
}
```

**Comment by jerryshao** [comment]
CC @yiheng

**Comment by yiheng** [comment]
LGTM

File: server/src/main/scala/org/apache/livy/sessions/SessionManager.scala [hunk 2]
```diff
@@ -96,14 +96,18 @@ class SessionManager[S <: Session, R <: RecoveryMetadata : ClassTag](
     synchronized {
       session.name.foreach { sessionName =>
         if (sessionsByName.contains(sessionName)) {
-          throw new IllegalArgumentException(s"Duplicate session name: ${session.name}")
+          val errMsg = s"Duplicate session name: ${session.name} for session ${session.id}"
+          error(errMsg)
+          session.stop()
```
Comment (yiheng) [inline server/src/main/scala/org/apache/livy/sessions/SessionManager.scala]: It's a little weird that we call `session.stop()` before `session.start()`...

Comment (yiheng) [reply server/src/main/scala/org/apache/livy/sessions/SessionManager.scala]: From the code, when the duplicate name exception is thrown, the `session.start()` has not been called. Please let me know if I miss something.

**Comment by jerryshao** [comment]
Reopen to test it.

File: server/src/main/scala/org/apache/livy/sessions/SessionManager.scala (hunk 2 above)
Comment (shanyu) [reply server/src/main/scala/org/apache/livy/sessions/SessionManager.scala]: Yeah it a little odd.

createSession() calls InteractiveSession.create(), which calls LivyClientBuilder().build(), which creates a LivyClient, then create an InteractiveSession with this client. At this time the separate process is running which creates a Yarn application. This is before session.start() is called.

But to clean this Yarn application, we need to call session.stop().

**Comment by jerryshao** [comment]
Reopen to test again.

**Pull request closed without merging**